[Event Hubs] Event Batch Enhancements

jsquire opened the pull request:
# Summary

The focus of these changes is to avoid paying the cost of serialization from `EventData` instances to AMQP messages multiple times, where possible.

# Details

Previously, events were serialized to AMQP format when `TryAdd` was called for the `EventDataBatch` in order to measure the event size, and then the `EventData` instance was cloned and held in the batch.  When calling `SendAsync`, the events held by the batch were again serialized to AMQP format for each attempt at publishing; when retries were needed, the full serialization pass was made against all batch events.

This approach was used for two reasons.  First, the AMQP transport tracks delivery information on the serialized message directly and prevents messages from being sent twice; in error scenarios, attempts to retry sending were failing because of the transport behavior.  Additionally, when idempotent retries are enabled, the metadata for sequencing events must be applied as part of the send operation, preventing eager serialization and caching.

The changes included herein work around the transport limitation by allowing previously serialized messages to be packaged in a new AMQP batch envelope for each attempt to publish.  The transport tracks delivery on the envelope, which is discarded in the event of retries and is rebuilt without needing to perform a full serialization pass over the batch events.  This removes the need to clone the source event in `TryAdd` and allows the batch to hold the serialized AMQP message, eliminating the need to perform serialization for batch events when `SendAsync` is called.

When idempotent retries are enabled, the idempotent metadata written to an event cannot be calculated eagerly and must be determined at the time `SendAsync` is called.  Because of this, the transport batch now holds responsibility for applying sequencing and other publisher properties to AMQP messages in batch scenarios.  For non-batch scenarios, the producer will continue to apply sequencing to the event prior to serialization.

# References and Related

- [Revisit the ability to cache AMQP messages when adding events to the AmqpEventBatch (#9569)](https://github.com/Azure/azure-sdk-for-net/issues/9569)

jsquire committed:
[Event Hubs] Event Batch Enhancements

The focus of these changes is to avoid paying the cost of serialization
from `EventData` instances to AMQP messages multiple times, where possible.

Previously, events were serialized to AMQP format when `TryAdd` was called
for the `EventDataBatch` in order to measure the event size, and then the
`EventData` instance was cloned and held in the batch.  When calling
`SendAsync`, the events held by the batch were again serialized to AMQP
format for each attempt at publishing; when retries were needed, the full
serialization pass was made against all batch events.

This approach was used for two reasons.  First, the AMQP transport tracks
delivery information on the serialized message directly and prevents
messages from being sent twice; in error scenarios, attempts to retry
sending were failing because of the transport behavior.  Additionally,
when idempotent retries are enabled, the metadata for sequencing events
must be applied as part of the send operation, preventing eager serialization
and caching.

For mainline scenarios, the changes included herein work around the
transport limitation by allowing previously serialized messages to be
packaged in a new AMQP batch envelope for each attempt to publish.  The
transport tracks delivery on the envelope, which is discarded in the
event of retries and is rebuilt without needing to perform a full
serialization pass over the batch events.  This removes the need to clone
the source event in `TryAdd` and allows the batch to hold the serialized
AMQP message, eliminating the need to perform serialization for batch
events when `SendAsync` is called.

When idempotent retries are enabled, many of the benefits are negated;
the idempotent metadata written to an event cannot be calculated eagerly
and must be determined at the time `SendAsync` is called.  Because of this,
the previous behavior of cloning `EventData` instances in `TryAdd` and
holding those instances in the batch is retained.   Additionally, a full
serialization pass over the batch attempts is neeed when `SendAsync` is
called.  Some efficiency is gained in retry scenarios, as the previously
serialized messages can be reused with a new AMQP envelope for each
publishing attempt rather than performing a full serialization pass against
the source events as had been previously done.

azure-sdk commented:
**API change check for `Azure.Messaging.EventHubs`**

API changes are not detected in this pull request for `Azure.Messaging.EventHubs`

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpProducer.cs:
```diff
@@ -454,23 +470,17 @@ public override async Task CloseAsync(CancellationToken cancellationToken)
-                        // again after the operation completes to provide best efforts in respecting it.
-
-                        EventHubsEventSource.Log.EventPublishStart(EventHubName, logPartition, operationId);
+                        using AmqpMessage batchMessage = MessageConverter.CreateBatchFromMessages(messages);
```
This approach requires converting each AMQP message into an `ArraySegment<byte>` to be packaged in the envelope.  For retry scenarios, this work is performed for each attempt.  The alternative would be to create the batch envelope once and then clone it for each send attempt.   That requires the first send attempt to pay for creation of the batch envelope creation and a clone, while retries pay only for the clone.

I chose to optimize for the golden path here - since a single send attempt is the majority outcome, by creating a new batch envelope during each attempt, the first send avoids paying for an unneeded clone.

danielmarbach commented:
@jsquire happy to look at it some time on the weekend. When do you need feedback the latest?

jsquire commented:
> @jsquire happy to look at it some time on the weekend. When do you need feedback the latest?

Awesome!  I'd like to start running things through prolonged stress tests the week of the 18th.   Even if I end up closing this PR out get some burn in during nightly test runs before you get a chance to give it a look, I'd love to incorporate your feedback in another set of changes.

danielmarbach commented on sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpMessageConverter.cs:
```diff
@@ -295,16 +295,21 @@ public virtual PartitionProperties CreatePartitionPropertiesFromResponse(AmqpMes
                 {
                     case List<AmqpMessage> messageList:
-                        batchEnvelope = messageList[0];
+                        batchEnvelope = messageList[0].Clone();
```
It seems it does only a shallow copy. Is that a problem?

https://github.com/Azure/azure-amqp/blob/master/src/AmqpMessage.cs#L466-L468

danielmarbach commented on sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpProducer.cs:
```diff
@@ -203,8 +203,27 @@ internal class AmqpProducer : TransportProducer
             Argument.AssertNotClosed(_closed, nameof(AmqpProducer));
             Argument.AssertNotClosed(ConnectionScope.IsDisposed, nameof(EventHubConnection));
 
-            AmqpMessage messageFactory() => MessageConverter.CreateBatchFromEvents(events, sendOptions?.PartitionKey);
```
Yehaa. This factory was ugly... :partying_face:

jsquire replied:
yeah, it was.  The pivot to "keep events because the AMQP messages are throwing on retries" was very late in the ship cycle.

danielmarbach commented on sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpProducer.cs:
```diff
@@ -311,7 +327,12 @@ internal class AmqpProducer : TransportProducer
+            // If idempotent publishing is enabled, a specialized transport batch is needed to
+            // support sequencing semantics during send operations.
+
+            return ((ActiveFeatures & TransportProducerFeatures.IdempotentPublishing) == 0)
```
This is nitpicky. I was just wondering if this should be abstracted into `AmqpEventBatch.Create(MessageConverter, options, TransportProducerFeatures, ActiveFeatures)` and then return the base.

I was immediately wondering why we access ActiveFeatures here and then also passing it to the batch creation.

jsquire replied:
Interesting thought.  I'll prototype that and see how it feels.  

The feature flags are intended to support a few things forthcoming that may need to be considered as part of measurement, conversion, or sending.   Currently, it's just the idempotent coarse-grained flag.

jsquire replied:
I like it; centralizes the concern.   I'll push this up shortly.

danielmarbach commented on sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/IdempotentAmqpEventBatch.cs:
```diff
@@ -0,0 +1,206 @@
+    ///   using an AMQP-based transport appropriate for idempotent publishing.
+    /// </summary>
+    ///
+    internal class IdempotentAmqpEventBatch : AmqpTransportEventBatch
```
would it make sense to seal?

jsquire replied:
Can't do it in this case;  the model factory implementation is extending this.  😄

danielmarbach commented on sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/IdempotentAmqpEventBatch.cs:
```diff
@@ -0,0 +1,206 @@
+                // mutated.
+
+                _sizeBytes = size;
+                BatchEvents.Add(eventData.Clone());
```
See my previous comment about shallow vs deep clone

jsquire replied:
The intent, as you're likely aware, is to prevent mutating the batch size by manipulating the `EventData` source.   

Generally, we're trying to guard against setting the direct properties or adding additional members into the `ApplicationProperties` dictionary.   The properties are comprised of primitives and immutable types that we copy over to the clone.  The dictionary only supports primitive types and streams for serialization - the latter of which is a very unusual case.   _([ref](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpMessageConverter.cs#L942))_

In the case where someone put a stream in the application properties dictionary and is writing to it, and has idempotent retries enabled ... our abstraction leaks and manipulating the stream does mutate the batch.   I'd call it more like a mid-level copy - not fully deep but also not totally shallow.   It's not ideal, but it's been a decent balance for the mainstream case.

danielmarbach reviewed (commented):
Very clean and good separation of concerns. I think the design is intention revealing. 

I was wondering if it would be possible to separate the payload from the metadata since the AmqpMessage allows getting the payload and then maybe it would be possible to reconstruct the AmqpMessage with the payload and fresh metadata. With that, you wouldn't have to pay the price of serializing the payload again and again on the idempotent path. But I'm not deep enough into all the mechanics here, so this might just be nonsensical gibberish what I'm writing here.

jsquire committed:
Refactoring to use a factory method for creating the correct AmqpTransportEventBatch subtype

jsquire committed:
Fixing minor formatting error

jsquire commented:
> I was wondering if it would be possible to separate the payload from the metadata since the AmqpMessage allows getting the payload and then maybe it would be possible to reconstruct the AmqpMessage with the payload and fresh metadata. With that, you wouldn't have to pay the price of serializing the payload again and again on the idempotent path. But I'm not deep enough into all the mechanics here, so this might just be nonsensical gibberish what I'm writing here.

I'm not sure that I'm following.   I think what you're suggesting is to shift some of the logic for sequencing to the transport batch and manipulate it at the AMQP message level.    Something like:

```csharp
// Producer.SendIdempotent

var publishingState = GetPublishingState(partitionId);

try
{
    var lastBatchSequenceNumber= batch.ApplyPublishingState(publishingState);

    await transportProducer.SendAsync(batch, ...);
    publishingState.LastSequenceNumber = lastBatchSequenceNumber;
}
catch
{
    batch.ResetPublishingState();
}
```

Where internally the batch would do something like:

```csharp
public class Batch
{
    public long ApplyPublishingState(PartitionPublishingState state)
    {
        var sequenceNumber = GetNextSequenceNumber(state.LastSequenceNumber);

        foreach (var message in _batchMessages)
        {
              message.MessageAnnotations.Map[AmqpProperty.ProducerSequenceNumber] = sequenceNumber;
              sequenceNumber = GetNextSequenceNumber(sequenceNumber);

              message.MessageAnnotations.Map[AmqpProperty.ProducerGroupId] = stateProducerGroupId;
              message.MessageAnnotations.Map[AmqpProperty.ProducerOwnerLevel] = state.ProducerOwnerLevel;
        }
    }

    public long ResetPublishingState()
    {
        foreach (var message in _batchMessages)
        {
              message.MessageAnnotations.Map.TryRemoveValue<int?>(AmqpProperty.ProducerSequenceNumber, out _);
              message.MessageAnnotations.Map.TryRemoveValue<long?>(AmqpProperty.ProducerGroupId, out _);
              message.MessageAnnotations.Map.TryRemoveValue<short?>(AmqpProperty.ProducerOwnerLevel,  out _);
        }
    }
}
```
_(I'd probably do that in the message converter, but close enough for illustration....)_


Am I close?

danielmarbach commented:
Yep exactly. It was just a wild thought and I haven't validated if that would even work or give us much.

jsquire committed:
WIP | Tests pass

jsquire committed:
Refactoring idempotent implementation

This allows AMQP messages to be cached in idempotent scenarios, with
responsibility for applying sequencing shifted to the batch rather than
being purview of the producer.

jsquire commented:
> It was just a wild thought and I haven't validated if that would even work or give us much.

Not so wild; its a great suggestion that I didn't think of.   I refactored for this approach, which let me remove a lot of the idempotent-specific infrastructure.   I think it's a huge improvement to what was here.   Thanks!

jsquire commented:
/azp run net - eventhub - tests

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.sln:
```diff
@@ -35,14 +35,6 @@ Global
 		Release|Any CPU = Release|Any CPU
 	EndGlobalSection
 	GlobalSection(ProjectConfigurationPlatforms) = postSolution
-		{6221D9A4-8563-4B8B-86B2-73585D54F840}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
```
I do not know what caused this, but things seem happy...

danielmarbach replied:
:magic_wand:

danielmarbach commented:
This looks nice! :partying_face: :balloon:

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpEventBatch.cs:
```diff
@@ -123,43 +126,26 @@ public override bool TryAdd(EventData eventData)
-            // Reserve space for producer-owned fields that correspond to special
-            // features, if enabled.
-
-            if ((ActiveFeatures & TransportProducerFeatures.IdempotentPublishing) != 0)
```
I forgot to add idempotent property reservation back in after collapsing back to a single batch.     Will fix.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpEventBatch.cs:
```diff
@@ -183,12 +174,59 @@ public override void Clear()
+                return lastSequenceNumber;
+            }
+
+            _startingSequenceNumber = NextSequence(lastSequenceNumber);
```
I wanted to avoid embedding an `if` statement in the `foreach` loop, so I'm paying the cost of the unchecked increment/rollover twice rather than the conditional for each tick.

jsquire replied:
If felt it was cleaner to do this outside the loop and pay for an extra unchecked increment/rollover rather than have a conditional for each tick of the loop just to set this on the first tick.

jsquire committed:
Adding idempotent property size calculation back into the transport batch.

jsquire committed:
Removing the ability to request a collection of EventData from the batch in the model factory to conform behavior to the actual type.

christothes commented on sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClient.cs:
```diff
@@ -1111,46 +1110,33 @@ public virtual async ValueTask DisposeAsync()
-                        eventData.PendingProducerGroupId = partitionState.ProducerGroupId;
-                        eventData.PendingProducerOwnerLevel = partitionState.OwnerLevel;
-                    }
+                    var lastSequence = eventBatch.ApplyBatchSequencing(partitionState.LastPublishedSequenceNumber.Value, partitionState.ProducerGroupId, partitionState.OwnerLevel);
```
is `partitionState.LastPublishedSequenceNumber.HasValue` guaranteed to be true here?

jsquire replied:
It is.   The service sets a value for it when the link is opened if the idempotent publishing capability is set.  That happens via the `InitializePartitionStateAsync` call on L1108.   If the service can't initialize the link properties, an exception flow is triggered.

danielmarbach commented:
Looking forward to seeing the impact of this change with some nice profiler screenshots :D
